### PR TITLE
Update requirements for share validation

### LINKED ISSUE #408: VDAF security analysis: Requirements for report processing
The VDAF analysis [posted on the CFRG list](https://mailarchive.ietf.org/arch/msg/cfrg/-errBjFRvCqi7KuAxoZwH6iY4Vc/) suggests the following restrictions is needed:

> Poplar1: Applications MUST ensure that reports are aggregated at most once
> per-level. Repeated usage of the authenticator at a given level may lead to
> subtle attacks on privacy. In particular, the current
> "max_batch_query_count" mechanism of DAP [2, Section 4.5.6] is not
> sufficient.

By "level" we mean the level of the IDPF tree. The basic idea is that the client generates some randomness for each level of the tree, and it is not safe to use this randomness more than once.

Relatedly: 

> Prio3: Applications SHOULD ensure that each report is processed at most
> once. It's probably safe to allow re-processing, but (1) this is not
> captured in our model; (2) this gets hairy in terms of differential privacy
> (see issue https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/94); and (3)
> there doesn't seem to be a good reason to allow this. (It is explicitly
> disallowed in DAP.)

In other words, the safest bet for Prio3 is to make sure we don't use a proof more than once.

In general, VDAFs may need to restrict how a report is processed based on the aggregation parameters with which it was processed in the past.

### PR DESCRIPTION
**Pull request by simon-friedberger**
Fixes #408

### CODE REVIEW

**Comment by simon-friedberger** [comment]
Intended for #408

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1369]: I don't think a prescription like this suffices. This should be written in terms of some function exposed by VDAFs, so that DAP implementations can remain generic. So I believe we're blocked on a corresponding VDAF change here.

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1345]: Since the previous n which have been aggregated cannot be used anymore this is already covered. The only way to aggregate again is if there are another $threshold reports for the same time interval. In that case it's fine.

Maybe the phrasing is indeed the problem. Afaiu "is in a batch which has been collected" and "has never been aggregated" are contradictory. I assumed it meant "batch definition" i.e. "time interval".

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1369]: Agreed, this is supposed to be a bandaid. Until we have the VDAF change we have the only requirement that we know of covered here.

Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1369]: I actually think that @simon-friedberger is more or less on the right track here, however I agree we want to be generic. Here is where we have converged on the VDAF spec so far (https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/161, please feel free to offer feeedback there as well):

> A VDAF *MAY* specify the "validity scope" of each aggregation parameter. Protocols using a VDAF (like DAP) *MUST* ensure that for each validity scope, a report is aggregated at most once with an aggregation parameter of that validity scope. If no validity scope is specified, a report *MUST NOT* be aggregated more than once with a given aggregation parameter.

This provides what we need for the current VDAFs:

* For Poplar1 the validity scope is the IDPF tree level, which is included in the aggregation parameter; the language here seems sufficient to ensure we don't evaluate at the same level more than once.

* For Prio3 we have no validity scope, however the only valid aggregation parameter is the empty string. The language implies we don't aggregate the same report more than once [1].

So how does this translate into normative text for input share validation? Maybe something like this:

> Check if the input share was previously aggregated with the aggregation parameter's validity scope {{reference-to-definition-in-the-vdaf-draft}}. If so, the Aggregator *MUST* mark the input share as invalid with the error "XXX".

cc/ @schoppmp

[1] The attack we're considering here: We want to make sure the attacker can't query the same input/proof on multiple distinct query randomnesses. If we can ensure by some other means that both the nonce and verify_key can't change overtime, then this may not be strictly necessary. Since we haven't really settled the details here, making sure we aggregate each report at most once seems like a reasonable conservative choice.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1345]: > Since the previous n which have been aggregated cannot be used anymore this is already covered. The only way to aggregate again is if there are another $threshold reports for the same time interval. In that case it's fine.

Hmm, I think I'm missing something important here. Can you clarify? Is there language elsewhere that covers this?

So the requirement is: "once a batch has been defined, do not add any more reports to that batch".

Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1345]: Okay, so how about
```
   A report must never be used in more than one batch. If a report with the same report ID is contained in a batch    for which the `AggregateShare` has been sent the input share MUST be marked as invalid with the error `batch_collected`.
```
This makes the requirement that Chris stated explicit, and removes the somewhat contradictory statement about a batch which has been collected but contains a report which has not been aggregated.

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1369]: I agree with Tim, since this is VDAF specific it should be implemented on the VDAF side.
How about:
> Check if the input share was not previously aggregated with the aggregation parameter's validity scope by calling `VDAF.is_valid(report_id, agg_param)`. Otherwise, the Aggregator MUST mark the input share as invalid with the error "invalid_aggregation_parameter".

Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1369]: Something like that, yeah. However, `VDAF.is_valid()` would also need to take as input the the agg params that have been used for report_id in the past.

Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1369]: @schoppmp Yes, that sounds good, so `validity_scope = VDAF.validity_scope(agg_param)` would extract parts of the agg_param (or otherwise construct) an opaque bitstring and DAP ensures that (report_id, validity_scope) has never been aggregated before, right?

Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1369]: Yup, that sounds right to me.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [reply draft-ietf-ppm-dap.md:1345]: I don't think that phrasing is right, because to me it doesn't capture that you can't add a report to a batch once it has been collected. How about this:

```
1. Some VDAFs allow a batch to be collected multiple times with different aggregation parameters, but the aggregators must check that the same set of reports is being aggregated each time. A report may not be aggregated into a collection unless it was included in all previous collections for the batch. If this check fails, the input share MUST be marked as invalid with the error `batch_collected`. This prevents collectors from learning anything about small numbers of reports that are uploaded between two collections of a batch.
```

Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1345]: I was trying to clarify what "once it has been collected" means with "for which the `AggregateShare` has been sent". Since the helper cannot tell if collection has actually finished (and if there is pro-active aggregation maybe even if it has started) it can only check if it has sent the `AggregateShare` to the Leader.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [reply draft-ietf-ppm-dap.md:1369]: IIUC we want to write this section in terms of the interface being added in https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/161

**Comment by simon-friedberger** [comment]
Update with comments so far.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1366]: I would say the agg param validity scope check should be made for all VDAFs. Where we landed in 04 is to require that each VDAF specify a method for checking it: https://stackoverflow.com/questions/868568/what-do-the-terms-cpu-bound-and-i-o-bound-mean

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1373]: API is now `Vdaf.is_valid(agg_param: AggParam, previous_agg_params: Vec[AggParam]) -> Bool`. What we need to do is something like: "Look up all of the aggregation parameters with which the report was previously aggregated; if `Vdaf.is_valid(agg_param, previous_agg_params)` is false, then the input share MUST be marked ...."

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1352]: I think this check should go last, since it's kind of a "catch all": I.e., if the Aggregator did not have enough information to perform even one of the previous checks, then we have to invalidate the report.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1362]: nit: Wrap paragraphs at 80 characters.

**Review by cjpatton — CHANGES_REQUESTED** [review]

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [inline draft-ietf-ppm-dap.md:1358]:
```suggestion
   previous collections for the batch (checking that no reports are omitted
   from a batch is covered in {{batch-validation}}).
```

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1358]: These seem like two sentences to me so I'm not sure why we would remove the split. I'm removing the parentheses instead.

File: draft-ietf-ppm-dap.md
Comment (simon-friedberger) [reply draft-ietf-ppm-dap.md:1362]: I am a fan of fuzzy 80 but happy to fix. These all show as 80 or less for me, though.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1388]: nit: This ought to minimize the size of the patch.
```suggestion
1. Finally, if an Aggregator cannot determine if an input share is valid, it
```

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1348]: nit (grammar): Most of the items in this list follow the template: "Check if <SOMETHING BAD HAPPNES>." Can we change this to something like "Check if the aggregation parameter violates the validity scope of the report."

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1369]:
```suggestion
   previous collections for the batch. Checking that no reports are omitted
```

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [inline draft-ietf-ppm-dap.md:1376]: nit: IMO this doesn't need to be clarified here.
```suggestion
```

**Review by cjpatton — COMMENTED** [review]
Looking good, just a few minor comments left.

File: draft-ietf-ppm-dap.md
Comment (tgeoghegan) [reply draft-ietf-ppm-dap.md:1376]: I agree with Chris, but also agree with Simon that DAP should be more clear about the lifecycle or state machine of various objects in the protocol, because otherwise it's kinda hard for implementors to figure out stuff like when a helper aggregate share is considered to be collected. I'd like to have a section in DAP where we discuss all that, so that then other parts of the protocol like this could refer to it, and then it'd be clear just what we mean when we say that a report is aggregated or a batch is collected. We talked about that a bit in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/384. In any case, I certainly don't want to solve that problem in this PR. Maybe we could take Simon's text and include a TODO referencing #384?

**Review by tgeoghegan — APPROVED** [review]
LGTM, modulo Chris' observations.

File: draft-ietf-ppm-dap.md
Comment (cjpatton) [reply draft-ietf-ppm-dap.md:1348]: Yeah make sense :)

**Comment by simon-friedberger** [comment]
I have two more concerns:
1. We have `report_too_early` do we want to add `report_too_late`? This would cover the "validation state is no longer available" which we currently use as an example of `report_dropped`. It would be clearer to define a lifetime for the state and emit `report_too_late` when it has expired.
2. I think we should maybe change the names. `batch_redefined` would be more accurate than `batch_collected`. Or maybe `batch_overlap`. Same for `report_replayed`, this might be better as `agg_param_invalid` now.

**Comment by cjpatton** [comment]
>     1. We have `report_too_early` do we want to add `report_too_late`? This would cover the "validation state is no longer available" which we currently use as an example of `report_dropped`. It would be clearer to define a lifetime for the state and emit `report_too_late` when it has expired.
>     2. I think we should maybe change the names. `batch_redefined` would be more accurate than `batch_collected`. Or maybe `batch_overlap`. Same for `report_replayed`, this might be better as `agg_param_invalid` now.

I like these suggestions, but they should be discussed in a different PR. Note that  @tgeoghegan has in the past wanted to err on the side of not adding too many error variants.

`batch_overlap` seems better to me :)

**Review by cjpatton — APPROVED** [review]

**Comment by cjpatton** [comment]
@simon-friedberger please squash these changes into a single commit, then I'll merge it.

**Comment by simon-friedberger** [comment]
> @simon-friedberger please squash these changes into a single commit, then I'll merge it.

Didn't I already do that?

**Pull request merged**
